### ISSUE TITLE
Problem with driver.executeScript

### ISSUE THREAD

**Issue by abahdanovich-rtbh** [WaterfallEngineering/selenium-node-webdriver#2]
I'm trying to execute `examples/hello.js` example with selenium-server/chromedriver. It looks like fine, but at the end it finishes with no results. It looks like

```
return driver.executeScript(function () {
```

returns deferred object (why?), which, I suppose, must be resolved before next `then` block:

```
results.forEach(function (result) {
```

but `result` variable is an empty array.
When I try to debug that deferred object:

```
results = driver.executeScript(.....
result.then(function(r) { console.log(r); })
```

I get correct results.
When I use phantomjs, everething goes the same way except  that the last `then(function (results)` gets correct array of values and everything works fine.

**Comment by lawnsea** [comment]
Are you trying to run the example in `selenium-server/chromedriver` or are you trying to drive a Chrome server from Node?

**Comment by abahdanovich-rtbh** [comment]
I am trying the following:
Run selenium-server-standalone.jar instance which drives chrome via
chromedriver. Next, I run the google search the example and I can see
chrome running and doing the search, but at the same end, no results appear
in console.
PS. When i debug the example in node-inspector, the results do appear in
console sometimes (it looks like a race condition).
7 mar 2013 09:00, "Lon Ingram" notifications@github.com napisał(a):

**Comment by lawnsea** [comment]
I see. I have not tested this with any server other than Phantom 1.8, so I'm not surprised there are some hiccups.

Try this script and see if it resolves the race:

``` javascript
var webdriver = require('selenium-node-webdriver');

webdriver().
  then(function (driver) {
    driver.get('http://www.google.com').
      then(function () {
        return driver.
          findElement(driver.webdriver.By.name('q')).
          sendKeys('webdriver');
      }).
      then(function () {
        return driver.
          findElement(driver.webdriver.By.name('btnG')).click();
      }).
      then(function () {
        // wait for results to appear with exponential backoff
        // give up when next delay would be 2s
        return driver.executeAsyncScript(function (callback) {
          var delay = 10;

          function getResults() {
            var results = document.querySelectorAll('h3.r');

            if (results.length > 0) {
              callback(Array.prototype.slice.call(results).
                       map(function (result) {
                return result.textContent;
              }));
            } else if (delay < 2000) {
              delay *= 2;
              setTimeout(getResults, delay);
            }
          }

          getResults();
        });
      }).
      then(function (results) {
        results.forEach(function (result) {
          console.log(result);
        });
        driver.quit();
      });
  });
```

**Comment by abahdanovich-rtbh** [comment]
Thank you for reply.
This script ends with an error:

```
Timed out waiting for asyncrhonous script result after 3 ms (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 32 milliseconds
```

However, when I increase the limit, it works ok..

```
driver.manage().timeouts().setScriptTimeout(5000);
```

So, it looks like `h3` headers are not ready immediately... Chrome makes about 6-7 tries before it fulfills the `results.length > 0` condition...

Is there more elegant way to do the same, maybe something like `driver.wait(function(){ results.length > 0 })`?

**Comment by lawnsea** [comment]
Sorry for dropping out of this thread. Yes, I think waiting on results would be better than a timeout.

**Comment by abahdanovich-rtbh** [comment]
Ok, thank you for reply
